### ISSUE TITLE
Allow git repositories to live on a different server

### ISSUE THREAD

**Issue by embray** [sagemath/sage_trac_plugin#9]
Currently the plugin generally assumes that the 'main' repositories served by cgit / gitolite live on the same server as the Trac server.  I see no reason this needs to be the case though.  It might be useful at some point to fix that, but not a high priority.

**Comment by ohanar** [comment]
I don't see any good way to really do this decoupling. A few questions come to mind:
1. How would ssh keys be managed? I think the current way of using trac as a gateway to adding ssh keys makes a lot of sense.
2. Who would be responsible for making unrefed merges, the trac server or the git server? If it is the git server, then you would probably have to setup a separate daemon that the trac plugin could send requests for unrefed merges. If it is the trac server, the operation would become much more expensive because you are no longer locally cloning, moreover, you would have to figure out a way to push the unrefed commits (which wasn't supported when I wrote the code 2 years ago and is why the commit tree is manually written to the repo -- I don't know what the situation is now).

**Comment by vbraun** [comment]
I agree that its a significant effort, there would have to be some demon doing the checkout / merge / report on the git server. The patchbot does that basically already so maybe one could steal some code there.

But the benefits of decoupling are also evident, git operations can get very IO heavy and running them on the same file system as the trac server is obviously bad for performance.

**Comment by embray** [comment]
> How would ssh keys be managed? I think the current way of using trac as a gateway to adding ssh keys makes a lot of sense.

As it is gitolite stores the SSH keys and connection with gitolite is done over SSH (almost by requirement) so there's no reason for it to be on the same machine.  The UI for SSH key management can still be through Trac, but on the backend the Trac server is just committing to a local clone of the gitolite-admin repo, and then pushing to the main gitolite server (I'm actually in the process of updating the SSH key management aspect of the plugin to make this a little clearer).

> Who would be responsible for making unrefed merges

I'm not sure what you mean by "unrefed merges" so most of the rest of that paragraph is lost on me I'm afraid...

True that pushing changes to a remote gitolite-admin repo _could_ be slow from the user's perspective if their POST requests hang on it.  Fortunately SSH key updating isn't that frequent for most users.  If it does end up being an issue I can have the push happen asynchronously and display a flash message to the user if something went wrong.

Actually it might be best to have a separate process outside the webserver handling that, so that you don't have multiple server processes contending for control over a git repository.  As it is I add directory locks.

The current plugin is not better.  It just has a clone of the gitolite-admin repo in /home/www-data for some reason, and when the user submits the SSH key form it calls a hacky `gitolite-update` script directly in the request handler which pulls/pushes.

**Comment by embray** [comment]
> But the benefits of decoupling are also evident, git operations can get very IO heavy and running them on the same file system as the trac server is obviously bad for performance.

Yep, I actually wondered earlier this morning that the cause of the Trac server performance issues of the last couple days could be due to requests hanging on the git repo somehow.  Now I'm not sure that's the case, but it is a possibility.

**Comment by ohanar** [comment]
By "unrefed merges" I mean merges that are not in the history of any branch or tag (commits of this sort are ephemeral and git will garbage collect them). These are how the diffs are computed for a branch on a ticket.

**Comment by embray** [comment]
Got it, I realized that after the fact.  I was making the mistake of thinking your two points were both connected to SSH keys (in part since that's where my head was at at the time).
